### ISSUE TITLE
Provide a decorator/utility that can send a gRPC response.

### ISSUE THREAD

**Issue by trustin** [line/armeria#2607]
Sometimes, a user wants to write an Armeria-level decorator (i.e. `SimpleDecoratingHttpService`) which intercepts a gRPC request and sends a pre-defined gRPC response instead of delegating to the actual `GrpcService`.

This can be useful when you want to send a gRPC response for authentication failure, throttled requests, etc.

We could provide some serialization utility which allows a user to serialize a single gRPC response into an `HttpResponse`, for example. I believe this is doable somehow by using `ArmeriaMessageFramer` but it would be nice if it's simpler and more convenient.

Similarly, we might want to provide a way to deserialize a gRPC `HttpResponse` into a `StreamMessage` of gRPC messages and others.

/cc @okue

**Comment by anuraaga** [comment]
I think gRPC interceptors already handle most cases so it'd be a lot of duplicate functionality with upstream with questionable gain.

I think `RetryStrategy` wouldn't be handled by upstream utilities though so there's probably something to help there but better at the request log level?

**Comment by trustin** [comment]
Would you mind listing the upstream utilities which might be useful to the users, especially the counterparts of `AuthService` and `ThrottlingService`?

**Comment by anuraaga** [comment]
It's true that upstream doesn't provide such business logic utilities. I'm not sure if this issue is about a layer for decoration, which is the server interceptor, or business logic. For reference, there's this

https://grpc.github.io/grpc-java/javadoc/io/grpc/ServerInterceptor.html

But I also don't see how to adapt classes like `AuthService` to gRPC since they work at the HTTP level. We could extract into a class that works with generic interfaces / objects, similar to brave, but that's a different topic I think.

> which intercepts a gRPC request and sends a pre-defined gRPC response instead of delegating to the actual GrpcService.

I'm mostly looking at this. All I can imagine if trying to reuse `AuthService`, not to intercept a gRPC request (which is available with interceptors and no need to use auth service which works with headers) but to intercept an HTTP request, namely headers, and return a gRPC response. Mixing the HTTP and gRPC layer like this seems like a dangerous slope, for such services gRPC would much rather recommend sending a proto back in the HTTP trailers.

**Comment by anuraaga** [comment]
@trustin Any current thoughts on this? I think the discussion about per request timeout can show the danger of mixing up the http and gRPC layers and really recommend gRPC is left to the gRPC logic layer.

https://line-armeria.slack.com/archives/C1NGPBUH2/p1585521198025400?thread_ts=1585521198.025400&cid=C1NGPBUH2

**Comment by trustin** [comment]
> All I can imagine is trying to reuse AuthService, not to intercept a gRPC request (which is available with interceptors and no need to use auth service which works with headers) but to intercept an HTTP request, namely headers, and return a gRPC response. Mixing the HTTP and gRPC layer like this seems like a dangerous slope, for such services gRPC would much rather recommend sending a proto back in the HTTP trailers.

I agree that sending a gRPC response at HTTP layer can be a place of mistake, but I'm not sure if the risk is that high for a simple gRPC responses which will be returned immediately. It can be detected and fixed easily. The point is that we allow a user to do that when he or she really has to do, not recommending to do it even when he or she doesn't need to do it.

> I think the discussion about per request timeout can show the danger of mixing up the http and gRPC layers and really recommend gRPC is left to the gRPC logic layer.

It seems to me that request timeout is a bigger problem that requires a different fix - our gRPC client does not respect the timeout set in the context and a user has to configure the timeout in two places.

**Comment by anuraaga** [comment]
I guess having a tool might be ok. Hopefully it's implemented when someone really needs it :)

> our gRPC client does not respect the timeout set in the context

The timeout in the context is the only implementation of timeout we have - we copy a gRPC deadline if it was set but otherwise just use it, no need to set in two places. The issue is grpc protocol also propagates timeout value to the server as a header - if a user changes the timeout using context after the HTTP request is sent, the client will use that new timeout, but naturally a new header won't be sent to the server.

**Comment by anuraaga** [comment]
> The point is that we allow a user to do that when he or she really has to do

Thought a bit more and less convinced this would ever happen. The user has to return the right response type for the method being handled, isn't this too strange?

**Comment by trustin** [comment]
> Thought a bit more and less convinced this would ever happen. The user has to return the right response type for the method being handled, isn't this too strange?

Could you elaborate, @okue?

**Comment by trustin** [comment]
> The user has to return the right response type for the method being handled

Maybe sending an exception response?

Besides this, I guess it will be useful for those who extend `AbstractUnaryGrpcService`.

**Comment by anuraaga** [comment]
> Maybe sending an exception response?

A exception thrown by a decorator should already be translated into an error response correctly - if not it's a bug and should be able to fix it :)

> Besides this, I guess it will be useful for those who extend AbstractUnaryGrpcService.

That class exists go handle gRPC protocol, including framing and exception conversion, so user doesn't have to do anything special here.

**Comment by trustin** [comment]
> A exception thrown by a decorator should already be translated into an error response correctly - if not it's a bug and should be able to fix it :)

I meant a decorator sending a gRPC response with a `StatusRuntimeException` + some metadata.

**Comment by anuraaga** [comment]
Yeah `StatusRuntimeException` can be constructed with metadata and it will be in the trailers as well as the exception caught in the client. Throwing the exception, or completing the http response with it, should work the same as if throwing the exception in a gRPC stub.

**Comment by trustin** [comment]
Hmm, but the exception raised by a decorator will not be handled like that because it's `GrpcService` who does the translation, right?

**Comment by anuraaga** [comment]
Oops for some reason was thinking of the client, not server - things are free on client since gRPC runs before decorators.

I wonder if the best pattern on server is to abort the request with the status exception and continue to delegate. If the decorator completely intercepts, none of the gRPC logic runs, including possibly important ones like server  call interceptors.

**Comment by trustin** [comment]
> If the decorator completely intercepts, none of the gRPC logic runs, including possibly important ones like server call interceptors.

That may be the intention of a user. A decorator or an interceptor is supposed to do that by design.

**Comment by anuraaga** [comment]
I guess so though it seems easy to shoot your foot - if we add a strong note pointing to interceptors to such logic it's not so bad though.

As an aside though, if the specific use case this time happens to be throttling I'd take a look at Netflix's library, hear good things about it.

https://github.com/Netflix/concurrency-limits/tree/master/concurrency-limits-grpc/src/main/java/com/netflix/concurrency/limits/grpc/server

**Comment by okue** [comment]
As Implementing gRPC server interceptor is a bit annoying to me, I implemented some logics (such as authentication, rate limit) with armeria decorators.

In error cases, my implementation returns the response like the following.

```kotlin
        HttpResponse.of(
            ResponseHeaders.builder()
                .endOfStream(true)
                .add(HttpHeaderNames.STATUS, "200")
                .add(HttpHeaderNames.CONTENT_TYPE, "application/grpc+proto")
                .apply {
                    GrpcTrailersUtil.addStatusMessageToTrailers(this, code, message)
                }
                .build()
        )
```

Possibly, I should use gRPC interceptor..?

**Comment by trustin** [comment]
I wonder if it's a better idea for us to provide an easy way to write a gRPC interceptor. gRPC-Java's interceptor API isn't the best API to implement simple logic like @okue's example... e.g.

```java
public class UnaryGrpcInterceptor implements ...<grpc interceptor type>... {
    public static UnaryGrpcInterceptor of(...<some function?>...) { ... }
    ...
}
```

(Sorry for a very vague example. I have no working memory on the interceptor API 😉 )

**Comment by okue** [comment]
Maybe, I should have implemented the logic at not the decorator but grpc-interceptor like this example.
https://github.com/saturnism/grpc-by-example-java/blob/master/metadata-context-example/src/main/java/com/example/grpc/server/JwtServerInterceptor.java#L51-L54

**Comment by okue** [comment]
BTW, I noticed `listener` is null during `ArmeriaServerCall.invokeOnCancel()` if `serverCall.close()` is called by `io.grpc.ServerInterceptor`.

https://github.com/line/armeria/blob/8a1281f1d2b827f516247b8f9eb6210f2a047efa/grpc/src/main/java/com/linecorp/armeria/server/grpc/ArmeriaServerCall.java#L492-L501

NPE is just ignored, because `Throwable` is caught and `closeCalled` is true in this case.
Is this expected?

**Comment by trustin** [comment]
I think it's not intended. Needs to be fixed 😭

**Comment by ikhoon** [comment]
> NPE is just ignored, because Throwable is caught and closeCalled is true in this case.

Let me take a look.

**Comment by octo47** [comment]
I would interested in some solution as well.
My case is I need async ServerInterceptor. Unfortunately ServerInterceptor is a synchronous and to make it async some significant effort required (seen various solutions around keeping queue of incoming requests and only pass them once auth is compete). Such solution is quite brittle.

On the other hand Armeria has a brilliant async infrastructure with decorators.

At the moment I test the solution where I use Armeria Authorizer to pre-auth user (i.e set AuthorizationResult in the ServiceRequestContext) and implement ServerInterceptor which will just make sure context has successful result.

Generally it works so far, unfortunately it scary to configure as you need to configure in two places: grpc and armeria.

Please, consider adding some form of async gRPC interceptors, even if they will not be gRPC compatible (as they are not anyway because of Context), but give a nice benefit over stock grpc to those who needs async interceptors.

**Comment by trustin** [comment]
Thanks, @octo47 for your feed back. I didn't know `ServerInterceptor` is synchronous 😱 (Just double checking, is it true, @anuraaga ?) If that's the case, it sounds like it makes a good sense to implement what's been proposed in the issue description.

**Comment by be-hase** [comment]
yes. ServerInteceptor is synchronous.
A similar problem has been pointed out in gRPC-kotlin.
https://github.com/grpc/grpc-kotlin/issues/223
